src/v/storage/disk_log_impl.cc

```diff
@@ -2567,25 +2578,31 @@ disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
           // (where dirty = base - 1). We don't want to remove it because
           // batches may be concurrently appended to it and we should keep them.
           return _segs.empty()
-                 || _segs.front()->offsets().get_base_offset()
-                      >= cfg.start_offset
-                 || _segs.front()->offsets().get_dirty_offset()
-                      >= cfg.start_offset;
+                 || keep_segment_after_prefix_truncate(
+                   _segs.front(), cfg.start_offset);
       },
-      [this] {
+      [this, prefix_truncate_offset = cfg.start_offset] {
+          // it is safe to capture the front segment pointer here. This
+          // operation is executed under the segment_rewrite_lock, we are
+          // guaranteed that no other operation will remove the segment from the
+          // segment list head (front).
           auto ptr = _segs.front();
-          // evict readers before trying to grab a write lock to prevent
-          // contention
           return _readers_cache->evict_segment_readers(ptr).then(
-            [this, ptr](readers_cache::range_lock_holder cache_lock) {
-                /**
-                 * If segment has outstanding locks wait for it to be unlocked
-                 * before prefixing truncating it, this way the prefix
-                 * truncation will not overlap with appends.
-                 */
+            [this, ptr, prefix_truncate_offset](
+              readers_cache::range_lock_holder cache_lock) {
                 return ptr->write_lock().then(
-                  [this, ptr, cache_lock = std::move(cache_lock)](
+                  [this,
+                   ptr,
+                   prefix_truncate_offset,
+                   cache_lock = std::move(cache_lock)](
                     ss::rwlock::holder lock_holder) {
+                      // after the lock is acquired, check if the segment is
+                      // still eligible for deletion as there might have been
+                      // concurrent appends
+                      if (keep_segment_after_prefix_truncate(
+                            ptr, prefix_truncate_offset)) {
+                          return ss::make_ready_future<>();
+                      }
```

Comment:
nit: assertion that checks if the `_segs` is not empty (there is a scheduling point, we have a lock but still)